zipline/pipeline/loaders/events.py

```diff
@@ -0,0 +1,154 @@
+        #  purposes.
+        self.events_by_sid = (
+            events_by_sid.copy()
+        )
+        dates = self.all_dates.values
+
+        for k, v in iteritems(events_by_sid):
+            # First, must convert to DataFrame.
+            if isinstance(v, pd.Series):
+                # If Series was passed, DateTime index is assumed.
+                self.events_by_sid[k] = pd.DataFrame(v)
+            elif isinstance(v, pd.DatetimeIndex):
+                if not infer_timestamps:
+                    raise ValueError(
+                        "Got DatetimeIndex for sid %d.\n"
+                        "Pass `infer_timestamps=True` to use the first date in"
+                        " `all_dates` as implicit timestamp." % k
+                    )
+                self.events_by_sid[k] = pd.DataFrame(v)
+                v.index = [dates[0]] * len(v)
+            # Already a DataFrame
+            elif isinstance(v, pd.DataFrame):
+                if TS_FIELD_NAME not in v.columns:
+                    if not infer_timestamps:
+                        raise ValueError(
+                            "Got DataFrame without a '%s' column for sid %d.\n"
+                            "Pass `infer_timestamps=True` to use the first "
+                            "date in `all_dates` as implicit timestamp." %
+                            (TS_FIELD_NAME, k)
+                        )
+                    self.events_by_sid[k] = v = v.copy()
+                    v.index = [dates[0]] * len(v)
+                else:
+                    self.events_by_sid[k] = v.set_index(TS_FIELD_NAME)
+            else:
+                raise ValueError("Data for sid %s must be in DataFrame, "
+                                 "Series, or DatetimeIndex." % k)
+            # Once data is in a DF, make sure columns are correct.
+            cols_except_ts = (set(v.columns.values) -
+                              {TS_FIELD_NAME} -
+                              {SID_FIELD_NAME})
+            # Check that all columns other than timestamp are as expected.
+            if cols_except_ts != expected_cols:
+                raise ValueError(
+                    "Expected columns %s for sid %s but got columns %s." %
+                    (expected_cols, k, v.columns.values)
+                )
+        self.dataset = dataset
+
+    def get_loader(self, column):
+        if column in self.dataset.columns:
+            return getattr(self, "%s_loader" % column.name)
+        raise ValueError("Don't know how to load column '%s'." % column)
+
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        return merge(
+            self.get_loader(column).load_adjusted_array(
+                [column], dates, assets, mask
+            )
+            for column in columns
+        )
+
+    def mk_date_series(self, date_field_name):
+        return {sid: pd.Series(index=event.index,
+                               data=np.array(event[date_field_name]))
```

Comment:
this is going to be making a lot of copies of the data. instead we can just take slices with:

```python
{sid: event_df[date_field_name] for sid, event_df in iteritems(self.events_by_sid)}`
```